codegen/src/api/mod.rs

```diff
@@ -175,15 +175,56 @@ impl RuntimeGenerator {
     ///
     /// Supported versions: v14 and v15.
     pub fn new(metadata: RuntimeMetadataPrefixed) -> Self {
-        let metadata = match metadata.1 {
+        let mut metadata = match metadata.1 {
             RuntimeMetadata::V14(v14) => metadata_v14_to_latest(v14),
             RuntimeMetadata::V15(v15) => v15,
             _ => panic!("Unsupported metadata version {:?}", metadata.1),
         };
 
+        Self::ensure_unique_type_paths(&mut metadata);
+
         RuntimeGenerator { metadata }
     }
 
+    /// Ensure that every unique type we'll be generating or referring to also has a
+    /// unique path, so that types with matching paths don't end up overwriting each other
+    /// in the codegen. We ignore any types with generics; Subxt actually endeavours to
+    /// de-duplicate those into single types with a generic.
+    fn ensure_unique_type_paths(metadata: &mut RuntimeMetadataV15) {
+        let mut visited_path_counts = std::collections::HashMap::<Vec<String>, usize>::new();
+        for ty in &mut metadata.types.types {
+            // Ignore types without a path (ie prelude types).
+            if ty.ty.path.namespace().is_empty() {
+                continue;
+            }
+
+            let has_valid_type_params = ty.ty.type_params.iter().any(|tp| tp.ty.is_some());
+
+            // Ignore types which have generic params that the type generation will use.
+            // Ordinarily we'd expect that any two types with identical paths must be parameterized
+            // in order to share the path. However scale-info doesn't understand all forms of generics
+            // properly I think (eg generics that have associated types that can differ), and so in
+            // those cases we need to fix the paths for Subxt to generate correct code.
+            if has_valid_type_params {
```

Comment:
Can't we just skip this check and modify duplicate the path the regardless to be on the safe side?

Maybe I don't understand the comment properly but I regard it as if it's generic is safe to have 
duplicate types because when those are instantiated it isn't an issue?

Would be cool to to have test for it :)

Reply:
If the path is same and the type has generics, `scale-info` already handles that case ok.

For example, in the metadata you might see types like `Vec<u8>`, `Vec<bool>` and `Vec<Foo>` mentioned. When we generate the code, we see that they all have the same path but don't care, so we just end up with a single `Vec<Something>` in the generated output. (I didn't realise this at first and indeed didn't have this check to begin with, but then the codegen ended up with `Vec1<T>`, `Vec2<T>`, `Vec3<T>` and so on for loads of types).

The problem we ran into here was that there were two types mentioned in the metadata, `pallet::Call` and `pallet::Call`. They actually had some different types in their variants, but didn't have any generics that the code gen could use, so appeared the same. Now, we spot that case and change one to `pallet::Call2` so that both get generated in the output and one doesn't end up overriding the other.

Yeah I can probably add a test to show that generics end up being output as just one type :)